New image state (and VK_KHR_separate_depth_stencil_layouts)

bjoeris opened the pull request:
## Description

This PR merges the `ImageLayouts` and `ImgRefs`, and additionally

- Adds support for the VK_KHR_separate_depth_stencil_layouts extension
- Adds correct image-subresource queue family tracking

Most of these changes are currently behind an off-by default compile time flag (RDOC_NEW_IMAGE_STATE).

I intend to do more testing, and I would like to add some more explanation for some of the changes here, but tests are passing (unit tests, python tests, and manual tests on games), and I don't expect to need any more big changes to this code. I wanted to get the PR rolling so that it can hopefully make it in time for the next release.

bjoeris committed:
Make queue family available in vkCmd* functions

This allows much simpler analysis of queue families. E.g. in
`vkCmdPipelineBarrier`, we can differentiate between queue family
release and acquire operations; we can also mark used subresources as
implicitly acquired by the command buffer's queue family.

Change-Id: I3e90ea1be5938781bdec675e69918b5f2bc49a49

bjoeris committed:
Add last present queue and semaphores to swapchain info

This is used for queue family tracking of presented images.

Change-Id: I64ac88a8e4b379bd947f56f5cf2c1db929e8c620

bjoeris committed:
Add additional fields to ImageInfo

Three new fields are added to `ImageInfo`
- `VkImageType imageType`: this is used to distinguish 2D and 3D images,
  to correctly handle access through 2D image views.
- `VkImageLayout initialLayout`: this is the `initialLayout` in which
  the image was created (`UNDEFINED` or `PREINITIALIZED`). This is used
  to correctly initialize `PREINITIALIZED` images (See: TODO).
- `VkSharingMode sharingMode`: This is needed to correctly generate
  queue family transfers, because we must *not* generate queue family
  transfers for images with `CONCURRENT` sharing mode.

This also refactors `SwapchainInfo` to use `ImageInfo`, rather than
duplicating some of `ImageInfo`s fields.

Change-Id: Ic42077c382e4055cb27af069045c46eacbd122cc

bjoeris committed:
Add ImageState

`ImageState` tracks the state of images and their subresources.
This functionality was previously split between the `ImageLayouts` and
`ImgRefs` classes.

Change-Id: I3242417dacf73fe07765f9bcfd449599e373e10d

bjoeris committed:
Add ImageSubresourceRange

`ImageSubresourceRange` is a common representation of subresource
ranges of Vulkan images, including `VkImageSubresourceRange` and
`VkImageSubresourceLayers`.

Additionally, `ImageSubresourceRange` can represent a range of depth
slices--this is used to model access to 3D `VkImage`s through 2D
`VkImageView`s, where the view array layers are mapped to the image's
depth slices.

Change-Id: Id08329b499a02817ccd3487bdbb9c1bb0f86262a

bjoeris committed:
Add ImageBarrierSequence

`ImageBarrierSequence` represents a sequence of batches of
`VkImageMemoryBarrier`s. The intended use case is setting image state,
which may involve a combination of layout transitions and queue family
transfers, and may require multiple barriers executed on different
queue families, in the correct order.

It is moderately expensive to execute a sequence of separate queue
submissions, with synchronization to prevent overlapping execution.
`ImageBarrierSequence` attempts to address this in two ways:
- `ImageBarrierSequence::Merge` allows separate barrier sequences (e.g.
  on separate images) to be merged, so that, e.g. the first batches from
  each of the barrier sequences are all run together.
- `InlineSetupImageBarriers`/`InlineCleanupImageBarriers` allow barriers
  to be inlined into an existing command buffer, when possible. E.g.
  to read the contents of an image, the barrier sequence may simply have
  a layout transition to `TRANSFER_SRC_OPTIMAL` (assuming the image is
  already on the correct queue family), and this can be inlined into the
  command buffer before executing the copy, avoiding the need for an
  extra queue submission.

Change-Id: I7f2edc650d9ff66871c9be5711789bfe33ca8c5e

bjoeris committed:
Added ImageState serialization

Change-Id: I87b2bc26bf1fc7d56b464ad2d67b6dbdc0c72ff6

baldurk commented on renderdoc/common/globalconfig.h:
```diff
@@ -135,6 +135,8 @@
 #define RDOC_WAYLAND OPTION_OFF
 #endif
 
+#define RDOC_NEW_IMAGE_STATE OPTION_OFF
```
Since we're going to enable this soon anyway and given that your testing has shown it to be reasonably stable already, it should be enabled now and remove the define entirely. We're not going to turn it off in future so it'll make the code much easier to read.

(Also for future not, globalconfig.h should be reserved for things that really are global - this is local to the vulkan driver so it can go in vk_common.h)

bjoeris replied:
Done. This is now on by default, and RDOC_NEW_IMAGE_STATE is removed.

baldurk commented on renderdoc/driver/vulkan/renderdoc_vulkan.vcxproj:
```diff
@@ -100,6 +100,7 @@
     </Link>
   </ItemDefinitionGroup>
   <ItemGroup>
+    <ClCompile Include="imagestate_tests.cpp" />
```
Should add this to CMakeLists.txt as well for completeness. Though the tests shouldn't need to be tested on all platforms it means that a linux developer can still run all the tests they can when making changes.

bjoeris replied:
Done

baldurk commented on renderdoc/driver/vulkan/vk_core.h:
```diff
@@ -438,6 +447,10 @@ class WrappedVulkan : public IFrameCapturer
 
+  void SubmitAndFlushImageStateBarriers(ImageBarrierSequence *barriers);
+  void InlineSetupImageBarriers(VkCommandBuffer cmd, ImageBarrierSequence *batches);
+  void InlineCleanupImageBarriers(VkCommandBuffer cmd, ImageBarrierSequence *batches);
```
These functions assume a non-NULL parameter so it's safer and cleaner to take a reference rather than a pointer.

bjoeris replied:
Done. ImageBarrierSequence is now always passed by reference.

baldurk commented on renderdoc/driver/vulkan/vk_serialise.cpp:
```diff
@@ -3762,6 +3762,77 @@ void DoSerialise(SerialiserType &ser, ImageLayouts &el)
+template <class SerialiserType>
+void DoSerialise(SerialiserType &ser, TaggedImageSubresourceState &el)
+{
+  uint32_t &newQueueFamilyIndex = el.state.newQueueFamilyIndex;
```
Serialise functions should reflect the real structure they are serialising. I think this is trying to overlap with serialisation of previous `ImageRegionState` objects? This isn't the way to do that - be explicit about what is being serialised and add a conversion step if needed, rather than trying to serialise in-place like this.

bjoeris replied:
I have updated the serialization code to more closely reflect the serialized structs. The remaining cases where the serialization code diverges from the struct are:

- Version check for fields added to ImageInfo
- Subresource state map is converted from/to an array when serializing ImageState

baldurk replied:
Yeh those cases are fine and there's no better way to do that.

baldurk commented on renderdoc/driver/vulkan/vk_resources.cpp:
```diff
@@ -3207,6 +3207,1292 @@ VkImageAspectFlags FormatImageAspects(VkFormat fmt)
     return VK_IMAGE_ASPECT_COLOR_BIT;
 }
 
+ImageSubresourceRange ImageInfo::FullRange() const
```
This file is already admittedly too long, but this is a lot of code to add on a (mostly) self-contained track. I'd rather move anything related to image range/state tracking into a new file - vk_image_layouts.cpp or vk_image_states.cpp.

bjoeris replied:
All the new code has been moved to a new file (vk_image_states.cpp)

baldurk commented on renderdoc/driver/vulkan/vk_resources.cpp:
```diff
@@ -3207,6 +3207,1292 @@ VkImageAspectFlags FormatImageAspects(VkFormat fmt)
+  if(baseMipLevel > imageLevelCount)
+  {
+    RDCWARN("baseMipLevel (%u) is greater than image levelCount (%u)", baseMipLevel, imageLevelCount);
+    RDCDUMP();
```
You should not be calling RDCDUMP manually like this. Instead this should be an error message - `RDCERR` already triggers a breakpoint conditionally if a debugger is attached, and otherwise continues on (which is the preferred behaviour in case the error message is wrong or we can somehow recover).

bjoeris replied:
Yes, these were leftover from debugging, and should not have been committed. They are now removed.

baldurk commented on renderdoc/driver/vulkan/vk_resources.cpp:
```diff
@@ -3207,6 +3207,1292 @@ VkImageAspectFlags FormatImageAspects(VkFormat fmt)
+  return base1 >= base2 && base1 + count1 <= base2 + count2;
+}
+
+bool ValidateLevelRange(uint32_t &baseMipLevel, uint32_t &levelCount, uint32_t imageLevelCount)
```
This validate functions are fine to leave but I think they should be disabled in release builds.

bjoeris replied:
Perhaps these functions should be renamed. They are validating the range, but also updating the range, both to handle `VK_REMAINING_*` values, and to try to prevent crashes if there is an incorrect range (possibly due to bad input from the application being captured).

baldurk replied:
That makes sense, sorry I missed that it was updating the range at the same time as error checking.

I'm trying to think if there's a better name for that than validate and I'm not coming up with good options. Perhaps `Sanitise`? Similar to `SanitiseOldImageLayout`, though for the `VK_REMAINING_*` values it's not really sanitisation. It at least conveys more clearly that this is expected to be manipulating inputs into expected ranges, so that even if it does have some checks for incorrect ranges you don't make the mistake I did of thinking it's read-only debugging checks.

I think it's fine if you want to leave it as-is. Past the initial review it should be less likely for a reader to make that mistake.

baldurk commented on renderdoc/driver/vulkan/vk_manager.h:
```diff
@@ -264,18 +264,31 @@ class VulkanResourceManager : public ResourceManager<VulkanResourceManagerConfig
+
+  template <typename SerialiserType>
+  void SerialiseImageStates2(SerialiserType &ser, std::map<ResourceId, LockingImageState> &states);
+  bool DeserialiseImageRefs2(ReadSerialiser &ser, std::map<ResourceId, LockingImageState> &states);
```
`Deserialise` already has a specific meaning within RenderDoc, for freeing any memory that needed to be allocated during serialisation of structs. I know the term is used in some places for the reading mode of serialisation, but in RenderDoc 'serialise' is bidirectional.

This `DeserialiseImageRefs2` replaces the reading mode of `Serialise_ImageRefs`, but I think instead of changing the function you should leave `Serialise_ImageRefs` as it is and replace the implementation inside the `if(IsReplayingAndReading())` call. You could assert that the serialiser is reading, to be safe, but that function doesn't do anything special on writing so it's already fine to use on reading only.

At the same time, though it's not #ifdef'd out in the header here, `SerialiseImageStates` is no longer referenced anywhere in the code, so you can rename `SerialiseImageStates2` -> `SerialiseImageStates`. It was useful while developing both in parallel I imagine but now it can be collapsed as `SerialiseImageStates2` already contains the necessary serialise-version checks to handle old data.

bjoeris replied:
This is now called `Serialise_ImageRefs` again. The signature changed slightly from the original `Serialise_ImageRefs`, accepting a instead of an rdcarray to populate. `Serialise_ImageRefs` is also no longer templated over the serialiser.

`SerialiseImageStates2` is now just `SerialiseImageStates`.

baldurk commented on renderdoc/driver/vulkan/vk_resources.h:
```diff
@@ -1112,6 +1152,524 @@ typedef BitFlagIterator<VkImageAspectFlagBits, VkImageAspectFlags, int32_t> Imag
+};
+
+template <typename Barrier>
+struct BarrierSequence
```
Does this need to be templated?

bjoeris replied:
This is templated to allow sequences of `VkBufferMemoryBarrier`s as well in the future. Buffer ranges have the same queue-family ownership semantics as images, and could similarly require multiple barriers executed on different queues to restore the correct state.

baldurk commented on renderdoc/driver/vulkan/vk_resources.h:
```diff
@@ -1112,6 +1152,524 @@ typedef BitFlagIterator<VkImageAspectFlagBits, VkImageAspectFlags, int32_t> Imag
+{
+  // batches[batchIndex][queueFamilyIndex] = array of barriers to submit to queueFamilyIndex as part
+  // of batchIndex
+  rdcarray<rdcarray<rdcarray<Barrier> > > batches;
```
It looks like the batch indices are fixed and small? If so do you think it would be better on average to avoid an allocation and have a fixed array here? It seems like we won't have that many barrier sequences around and they're almost always on the stack while we're temporarily processing one image, so I think the fixed array would be better.

In fact if we're only having these around temporarily I'd say it would be a good idea to have a fixed array of 3 queue families since there are no known devices with more than 3 queue families (graphics, compute, transfer). That could be asserted on at device creation time and just error out if more queue families are seen. If you want to be safe you could have a fallback to a dynamic array like this (since if it's not used, the nested rdcarray doesn't take much space). This might mean an intermediary struct to ease use as a parameter in `ExtractBatch` etc.

For the future this would be good to have as a specialisation of rdcarray, with an amount of fixed internal storage (similar to rdcstr's small string optimisation) and optionally allowing overflow to memory for the queue case. There are many other cases where a small stack array would be preferable to allocation (especially if it has an allocation fallback to handle edge cases).

bjoeris replied:
This has been changed to fixed arrays for the batch and queue family indices.

For now, I'm just asserting that the batch indices and queue family indices are valid in `BarrierSequnece::Add`.

baldurk commented on renderdoc/driver/vulkan/wrappers/vk_resource_funcs.cpp:
```diff
@@ -2129,11 +2203,29 @@ bool WrappedVulkan::Serialise_vkBindImageMemory2(SerialiserType &ser, VkDevice d
+        else
+        {
+          state->isMemoryBound = true;
+          state->boundMemory = id;
```
Typo here with the id - this is the id of the image, not the id of the memory.

bjoeris replied:
Fixed

baldurk commented on renderdoc/driver/vulkan/vk_resources.h:
```diff
@@ -1112,6 +1152,524 @@ typedef BitFlagIterator<VkImageAspectFlagBits, VkImageAspectFlags, int32_t> Imag
+  VkDeviceSize boundMemoryOffset = 0ull;
+  VkDeviceSize boundMemorySize = 0ull;
+  FrameRefType maxRefType = eFrameRef_None;
+  VkImage handle = VK_NULL_HANDLE;
```
Storing the unwrapped handle here makes me a little uneasy, I've been caught in the past by unwrapping (or forgetting to unwrap), such that I try to consistently store all handles as wrapped unless it's an object with the lifetime of one function, and unwrap at the last minute before passing into the API. How awkward would it be to store this handle wrapped still and only unwrap when pushing into barriers for submission?

If that would require too much gymnastics, please at least explicitly rename it to `unwrappedHandle`. Similarly with the barriers - if they are expected to contain unwrapped images then name them 'unwrappedBarriers'.

bjoeris replied:
I changed this to use a wrapped handle (and renamed it `wrappedHandle` to be extra clear).

This did require a change in `VulkanReplay::GetTextureData` that I'm not 100% sure of: there is a temporary image (`tmpImage`) that previously only had an unwrapped handle. An `ImageState` is created for `tmpImage`, to simplify the barrier creation code, and to parallel the `ImageState` operations being performed on `srcImage`. With the change to using wrapped handles, I had to create a new wrapped handle for `tmpImage` (using `GetResourceManager()->WrapResource()`).

bjoeris replied:
I can similarly change `ImageBarrierSequence` to use wrapped handles, and unwrap them immediately before recording.

baldurk replied:
If you think having `ImageBarrierSequence` use wrapped handles wouldn't be too intrusive with wraps and unwraps then that would be preferable. If it's going to be too messy and too much code churn then I'm fine with leaving it unwrapped, since now things are named explicitly unwrapped/wrapped.

baldurk commented on renderdoc/driver/vulkan/wrappers/vk_queue_funcs.cpp:
```diff
@@ -268,9 +268,14 @@ bool WrappedVulkan::Serialise_vkQueueSubmit(SerialiserType &ser, VkQueue queue,
           BakedCmdBufferInfo &cmdBufInfo = m_BakedCmdBufferInfo[cmd];
 
+#if ENABLED(RDOC_NEW_IMAGE_STATE)
+          // ImageState::Merge(m_ImageStates, m_BakedCmdBufferInfo[liveCmd].imageStates, false);
```
Stale commented out code here?

bjoeris replied:
Removed

baldurk commented on renderdoc/driver/vulkan/wrappers/vk_misc_funcs.cpp:
```diff
@@ -694,7 +699,16 @@ VkResult WrappedVulkan::vkCreateFramebuffer(VkDevice device,
+            if(state && state->GetImageInfo().extent.depth > 1)
+            {
+              record->imageAttachments[i].barrier.subresourceRange.baseArrayLayer = 0;
+              record->imageAttachments[i].barrier.subresourceRange.layerCount = 0;
```
Should `layerCount` still be 1 here or is this change deliberate? I'm not actually sure if layerCount is expected to be 1 or `VK_REMAINING_ARRAY_LAYERS` for 3D images, or if it is ignored entirely, but unless this needs to be 0 for something I think we should leave it as is for clarity. It looks a bit weird if nothing else.

bjoeris replied:
Yes, `layerConut` should be 1. Fixed.

baldurk commented on renderdoc/driver/vulkan/wrappers/vk_device_funcs.cpp:
```diff
@@ -1431,6 +1431,12 @@ bool WrappedVulkan::Serialise_vkCreateDevice(SerialiserType &ser, VkPhysicalDevi
+#if ENABLED(RDOC_NEW_IMAGE_STATE)
+      if(!strcmp(createInfo.ppEnabledExtensionNames[i],
+                 VK_KHR_SEPARATE_DEPTH_STENCIL_LAYOUTS_EXTENSION_NAME))
+        m_SeparateDepthStencilLayouts = true;
```
We already track enabled status of extensions in the `InstanceDeviceInfo` struct, see e.g. `m_pDriver->GetExtensions(GetRecord(m_Device)).ext_KHR_maintenance1`.

You'll need to add `VK_KHR_separate_depth_stencil_layouts` to `DeclExts()` - I notice that the extension isn't enabled/whitelisted in this change, is that deliberate? This bool won't ever be true until it is, since the applications can't enable the extension.

bjoeris replied:
I think this is the only comment I have not had time to address yet. I will switch this over to use `InstanceDeviceInfo` and look into the whitelisting next week.

bjoeris replied:
I've been looking into proper extension support (vs just supporting separate depth/stencil layouts in the image state tracking). I think I mostly have it working, and am currently working on a sample app that actually uses this extension so that I can verify this.

I think it might make sense to start a separate PR to actually add the extension support, along with the sample/test of the extension. I can revert this `m_SeparateDepthStencilLayouts` stuff in this PR.

baldurk commented on renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp:
```diff
@@ -3323,8 +3519,10 @@ bool WrappedVulkan::Serialise_vkCmdExecuteCommands(SerialiserType &ser, VkComman
+        // if we're replaying a range but not from the start, we are guaranteed to only be
+        // replaying
+        // one of our executed command buffers and doing it to an outside command buffer. The
+        // outer
```
Accidental reformatting on this comment - I've done this myself if I have clang-format on save and at one point saved with imbalanced braces so everything got indented one level too many.

bjoeris replied:
Fixed

baldurk commented on renderdoc/driver/vulkan/vk_initstate.cpp:
```diff
@@ -104,6 +106,13 @@ bool WrappedVulkan::Prepare_InitialState(WrappedVkRes *res)
+    LockedImageStateRef state = FindImageState(im->id);
+
+    // if the image has no memory bound, nothing is to be fetched
+    if(!state || !state->isMemoryBound)
```
This path doesn't have a check for the queue family being foreign/external - it's not possible to sync correctly with a foreign queue family as we need the application to transition, if an image is currently owned externally we have to bail and give it no initial contents (and most images that are owned by an external queue is something like a VR backbuffer that has been given to the compositor process - so we don't need initial contents for it).

bjoeris replied:
I added a check here--if the image has any foreign/external owned subresources, then no initial contents are saved.

Ideally/eventually, we probably want to save initial contents for the subresources that are not owned by foreign queues (and maybe also only save initial contents for those subresources that actually need it, once we go down that road)

baldurk replied:
Yeh in theory we could save any non-foreign subresources to be slightly more correct but I suspect (without proving this definitively) that almost all uses of foreign/external queues are either for resources with only one subresource, or they pass the whole resource. It may not be worth the risk of obscure bugs, I'd probably at least wait until someone has a non-theoretical need for that.

baldurk commented on renderdoc/driver/vulkan/vk_initstate.cpp:
```diff
@@ -1532,22 +1576,41 @@ void WrappedVulkan::Apply_InitialState(WrappedVkRes *live, const VkInitialConten
+      RDCWARN("No image state found for image %s", ToStr(id).c_str());
+      return;
+    }
+    // TODO: set `initialized`
```
This `TODO` seems to be `TODONE` :).

bjoeris replied:
Removed

baldurk commented on renderdoc/driver/vulkan/vk_resources.h:
```diff
@@ -1112,6 +1152,524 @@ typedef BitFlagIterator<VkImageAspectFlagBits, VkImageAspectFlags, int32_t> Imag
+
+DECLARE_REFLECTION_STRUCT(ImageSubresourceState);
+
+struct TaggedImageSubresourceState
```
I normally try to avoid naming bikeshedding, but on my first reading `TaggedFoo` is a Foo that has some tag/extra data. In this case it's an image subresource range that has a state, so if anything it'd be a `TaggedImageSubresourceRange`. Maybe `ImageSubresourceStateForRange` would be better (if a tad more verbose) ?

bjoeris replied:
This has been renamed to `ImageSubresourceStateForRange`

baldurk commented on renderdoc/driver/vulkan/vk_manager.cpp:
```diff
@@ -245,6 +245,155 @@ void VulkanResourceManager::MergeBarriers(rdcarray<rdcpair<ResourceId, ImageRegi
+      }
+    }
+  }
+}
```
The new serialise of image states doesn't have the fixup pass which tries to merge subresource states that are all identical. This wasn't a perfect solution by any means it was important - if an application is using large texture arrays and populating them, then the array layers and mips can get split - take an array that's 1000 layers 2D images with 8 mips each, with a full mip chain. Once split it means the subresource map has around 8k entries which causes a noticeable perf hit, but usually we don't actually need that because mips/slices are transitioned to TRANSFER_DST then back to SHADER_READ and end up all identical most of the time.

As a partial fix I merged the layouts at serialise time on the assumption that most frames you capture the splitting happened prior to the capture, and we can safely merge at load time and not have to worry about re-splitting. Ideally we should merge at a more regular frequency.

bjoeris replied:
I added `Unsplit` to ImageSubresourceMap`. This attempts to collapse the state each of the 4 subresource dimensions (aspect, level, layer, slice), unless it finds subresources with different states that would be collapsed into a single state.

I also added some unit tests for `Unsplit`.

For now, `Unsplit` is only called in `SerialiseImageStates`, mimicking the previous behaviour.

baldurk commented on renderdoc/driver/vulkan/vk_manager.cpp:
```diff
@@ -245,6 +245,155 @@ void VulkanResourceManager::MergeBarriers(rdcarray<rdcpair<ResourceId, ImageRegi
+      if(HasLiveResource(Image))
+      {
+        ResourceId liveid = GetLiveID(Image);
+        if(liveid != ResourceId())
```
This check is redundant here after already checking `HasLiveResource`.

bjoeris replied:
This was based on the same pattern in `SerialiseImageStates`, which checks both `HasLiveResource(Image)` and `liveid != ResourceId()`. It looks like `HasLiveResource` only checks that `Image` is in `m_LiveResourceMap`, but doesn't verify that the liveid is non-nil.

It seems plausible that `m_LiveResourceMap` should never have nil live IDs, so I'm happy to remove this secondary check if you think it's unnecessary.

baldurk replied:
Yeh it's a little messy in that there are multiple parallel maps expected to be updated in sync - so if there's a live resource in `m_LiveResourceMap` then it should also have a live ID. At the least if that is violated then the bug is not in this code.

I think `SerialiseImageStates` only does both checks because the code has changed and previously those two checks were entirely separate, it should be safe to drop the second if.

baldurk commented on renderdoc/driver/vulkan/vk_resources.h:
```diff
@@ -1112,6 +1152,524 @@ typedef BitFlagIterator<VkImageAspectFlagBits, VkImageAspectFlags, int32_t> Imag
+  inline static bool IsDepthSplit(uint16_t flags)
+  {
+    return (flags & (uint16_t)FlagBits::IsDepthSplit) != 0;
+  }
```
Can you clarify what depth splitting means here? It looks like it's accounting for 3D textures viewed as 2D arrays having different layouts for different depth slices, but that's not valid according to the vulkan spec. When doing 3D-as-array any barriers that implicitly refer to the image views, via render passes, always transition the whole resource:

> Automatic layout transitions apply to the entire image subresource attached to the framebuffer. If the attachment view is a 2D or 2D array view of a 3D image, even if the attachment view only refers to a subset of the slices of the selected mip level of the 3D image, automatic layout transitions apply to the entire subresource referenced which is the entire mip level in this case.

I think for our ref tracking we should do the same - if one slice is written, the whole image is considered written, etc. Tracking at slice-granularity doesn't seem worthwhile otherwise.

bjoeris replied:
Depth splitting is essentially the same behaviour that was previously implemented in `ImgRefs`. This allows us to track the read/write state independently for each depth slices.

It should not be possible for subresources which differ only in depth to ever have different layouts. I will make sure this is handled correctly for automatically applied layout transitions (TODO).

When the image state is used to generate barriers, subresources with non-zero `baseDepthSlice` are skipped, so a single barrier is generated for the entire depth range, whether or not the state is split by depth.

The motivation here is that finer grain read/write state should improve the reset performance. Even if we don't try to apply image initial contents at the depth slice level, there will still be a benefit in certain cases. E.g., if the application reads from depth slice 0, and then writes to depth slice 1, that does not require a reset. But if we drop the depth slice info, that becomes a read-before-write, which does require a reset.

baldurk replied:
Ah right I see, so it's split up for the benefit of reference tracking rather than for layout tracking.

I'm a little skeptical that there would be enough gain for this to be worth the added complexity, but admittedly I have not nearly as much profiling of initial states application as you have. Is this a scenario you've seen come up for real workloads? If that's the case, or if you have high confidence it's going to have a measurable benefit, I'm fine with this going in. I primarily raised the issue because I thought it was being added for correctness (since this is an odd quirk of the 3D/2D aliasing behaviour).

baldurk reviewed (changes requested):
As a general note this pull request is far too large. In future you need to find a way to split it into smaller chunks, even if they are committed disabled and non-compiling, so it can be reviewed in more reasonable chunks. Aiming for 3-4 commits and ~1000 lines changed is a good rule of thumb. Since I want to be sure that KHR_separate_depth_stencil_layouts lands soon in time to get tested before the next RenderDoc release I've reviewed it anyway, but future PRs like this will be rejected.

I think this change needs some more testing, at least with regard to opening captures from previous versions. I compiled with the option turned on and loaded a capture of the `VK_Simple_Triangle` demo program and got a number of errors (though the replay was OK). Please be sure to test with loading captures created on v1.5 from as many different programs as possible.

Otherwise I think the change overall is fine in structure, I've left some comments to be addressed. With the amount of code change I can't be sure I've caught everything but hopefully testing will catch anything else. In future we definitely need to expand the `VK_Image_Layouts` test to cover a number of edge cases handled here and ensure as much as possible that we validate cleanly and handle it correctly.

Again given this PR's complexity I'd like to ask that you please make all PR feedback as new commits initially, until the review has completed. Then it can be amended/rebased as desired. Otherwise it will be hard to track changes in amongst everything this size.

bjoeris commented:
Apologies for the unwieldy PR, and thanks for all the comments.

I did some more testing, focusing on opening v1.5 captures. I hacked the Python test scripts to use v1.5 captures with this version of replay, and the tests are all passing (except VK_Parameter_Zoo, VK_Vertex_Attr_Zoo and VK_Texture_Zoo, which were also failing with v1.5)

I will continue working through these comments and testing on other applications.

bjoeris commented:
I've created a separate branch for a follow-up PR to actually add the extension support: https://github.com/bjoeris/renderdoc/tree/pr-separate_layouts_ext

baldurk commented:
Thanks for the continued work on this over the end-of-year break, and for dealing with the large PR feedback. I think there's only a couple more issues left (the unresolved ones) and mostly are up to your judgement whether you need to change anything or not.

If you can please rebase the changes now into the final set of commits and then force-push any fixes after that, that would be great. I just wanted to have new commits for the bulk of the feedback otherwise it would be hard to tally up against comments. It's mostly up to you which commits get squashed back and which don't - I'd prefer that you squash any little 'fix this' or comment/formatting changes, but otherwise it's your choice whether you squash back changes based on PR feedback or not.

baldurk commented:
For that vk_manager.cpp failure - printing vulkan handles cross-platform is hard unfortunately. I'd say just drop the handle from the message since I don't expect there's anything we could do post mortem with a pointer value except see that it's non-NULL and/or plausible. If you want to leave it, you can use the `GetWrapped` function (even if it's not actually a wrapped handle, as long as you don't dereference it then it's just a cast).

bjoeris committed:
Add Unknown FrameRefType

This represents a (sub)resource for which no usage info is available.
It should be conservatively assumed that such a resource needs to be
re-initialized before each replay (and this behaviour is reflected in
`InitReq()`).

Change-Id: I12235a6cb1c4b2e3e21bed8834653ec6a3aea009

bjoeris committed:
Add LockingImageState

Change-Id: I8f1adc8617295d4a86067908ce8df85ea04e8365

bjoeris committed:
Add new image tracking to capture side

One notable high-level change is the behaviour of `StartFrameCapture`.
The previous implementation serialised the image state when
`StartFrameCapture` was called. The new implementation just sets the
"old" (beginnin go frame) image state (layouts/queue families) to be the
current state at this point; at `EndFrameCapture` the full image state
is serialised, including both the "old" (beginning of frame) and "new"
(end of frame).

Change-Id: I04c18c16a97e065c7a7d0edf058ba3f4803dcf45

bjoeris committed:
Add new image tracking to replay side

Change-Id: I67ef6b1f6fd9b84621ec6700a8260c24afb35a05

bjoeris committed:
Add SEPARATE_DEPTH_STENCIL_LAYOUTS support in new image state

Change-Id: I54458143cd075c0239f55ccb7e50916d725188ec
